module/zfs/spa_misc.c

```diff
@@ -1786,6 +1849,83 @@ spa_log_class(spa_t *spa)
 	return (spa->spa_log_class);
 }
 
+metaslab_class_t *
+spa_special_class(spa_t *spa)
+{
+	return (spa->spa_special_class);
+}
+
+metaslab_class_t *
+spa_dedup_class(spa_t *spa)
+{
+	return (spa->spa_dedup_class);
+}
+
+/*
+ * Locate an appropriate allocation class
+ */
+metaslab_class_t *
+spa_preferred_class(spa_t *spa, uint64_t size, dmu_object_type_t objtype,
+    uint_t level, uint_t special_smallblk)
+{
+	boolean_t has_special_class = spa->spa_special_class->mc_groups != 0;
```

Comment:
can we move these conditional assignments to after the DMU_OT_IS_ZIL() check? No need to evaluate for ZILs and they tend to be latency-sensitive